feat: Additional Features (#1425)
discordjs/guide: guide/additional-features/cooldowns.md

# Cooldowns

Spam is something you generally want to avoid, especially if one of commands require calls to other APIs or takes a bit of time to build/send.

This section assumes you followed the [Command Handling](/guide/creating-your-bot/command-handling.md) part.

First, add a cooldown property to your command. This will determine how long the user would have to wait (in seconds) before using the command again.

> ```js {4}

Reviewer: This number highlights the incorrect line.

const { SlashCommandBuilder } = require('discord.js');
module.exports = {
	cooldown: 5,
	data: new SlashCommandBuilder()
		.setName('ping')
		.setDescription('Replies with Pong!'),
	async execute(interaction) {
		// ...
	},
};
```

> In your main file, initialize a [Collection](/guide/additional-info//collections.md) to store cooldowns of commands

Reviewer: ```suggestion
In your main file, initialize a [Collection](/additional-info//collections.md) to store cooldowns of commands
```

> ```js {1}

Reviewer: I don't think the highlight is needed, especially if there is only one line of code.

client.cooldowns = new Collection();
```

> The key will be the command names, and the values will be Collections associating the user's id (key) to the last time (value) this specific user used this specific command. Overall the logical path to get a specific user's last usage of a specific command will be `cooldowns > command > user > timestamp`.

Reviewer: I think the word "specific" is used too much here. Maybe this could be reworded?

In your main file, add the following code:

```js {1,3-5,7-10,12-14}
const { cooldowns } = client;
if (!cooldowns.has(command.data.name)) {
	cooldowns.set(command.data.name, new Collection());
}
const now = Date.now();
const timestamps = cooldowns.get(command.data.name);
const defaultCooldownDuration = 3;
const cooldownAmount = (command.cooldown || defaultCooldownDuration) * 1000;
if (timestamps.has(interaction.user.id)) {
	// ...
}
try {
	// ...
} catch (error) {
	// ...
}

> ```

Reviewer: There should be new lines before and after the `if`s and `try...catch`. The highlights seem to be incorrect also.

You check if the `cooldowns` Collection already has an entry for the command being used. If this is not the case, you can add a new entry, where the value is initialized as an empty Collection. Next, create the following variables:

1. `now`: The current timestamp.

> 2. `timestamps`: A reference to the Collection of user's id and timestamp key/value pairs for the triggered command.

Reviewer: I believe this is correct:

```suggestion
2. `timestamps`: A reference to the Collection of user ids and timestamp key/value pairs for the triggered command.
```

Continuing with your current setup, this is the complete `if` statement:

```js {2,4-7}
if (timestamps.has(interaction.user.id)) {
	const expirationTime = timestamps.get(interaction.user.id) + cooldownAmount;
	if (now < expirationTime) {
		const timeLeft = Math.round(expirationTime / 1000);
		return interaction.reply({ content: `Please wait <t:${timeLeft}:R> more second(s) before reusing the \`${command.data.name}\` command.`, ephemeral: true });
	}
}
```

Since the `timestamps` Collection has the user's id as the key, you can use the `get()`method on it to get the value and sum it up with the `cooldownAmount` variable to get the correct expiration timestamp and further check to see if it's expired or not.

The previous user check serves as a precaution incase the user leaves the guild. You can now use the `setTimeout` method, which will allow you to execute a function after a specified amount of time and remove the timeout.

```js {5-6}
if (timestamps.has(interaction.user.id)) {
	// ...
}
timestamps.set(interaction.user.id, now);
setTimeout(() => timestamps.delete(interaction.user.id), cooldownAmount);

> ```

Reviewer: The highlight is wrong here (there is no 6th line of code). Add a space after the `if` too.

# Cooldowns

Spam is something you generally want to avoid, especially if one of your commands require calls to other APIs or takes a bit of time to build/send.

::: tip
This section assumes you followed the [Command Handling](/creating-your-bot/command-handling.md) part.
:::

First, add a cooldown property to your command. This will determine how long the user would have to wait (in seconds) before using the command again.

```js {4}
const { SlashCommandBuilder } = require('discord.js');

module.exports = {
	cooldown: 5,
	data: new SlashCommandBuilder()
		.setName('ping')
		.setDescription('Replies with Pong!'),
	async execute(interaction) {
		// ...
	},
};
```

> In your main file, initialize a [Collection](/additional-info//collections.md) to store cooldowns of commands

Reviewer 2: ```suggestion
In your main file, initialize a [Collection](/additional-info/collections.md) to store cooldowns of commands
```

```js
client.cooldowns = new Collection();
```

The key will be the command names, and the values will be Collections associating the user's id (key) to the last time (value) this user used this command. Overall the logical path to get a user's last usage of a command will be `cooldowns > command > user > timestamp`.

In your main file, add the following code:

```js {1,3-5,7-10,12-14}
const { cooldowns } = client;

if (!cooldowns.has(command.data.name)) {
	cooldowns.set(command.data.name, new Collection());
}

const now = Date.now();
const timestamps = cooldowns.get(command.data.name);
const defaultCooldownDuration = 3;

> const cooldownAmount = (command.cooldown || defaultCooldownDuration) * 1000;

Author: in the v12 guide it uses `||`

> const cooldownAmount = (command.cooldown || defaultCooldownDuration) * 1000;

Reviewer: Version 12 didn't support `??`, so yeah, that's why. I think it should change here.

	// ...
}
```

You check if the `cooldowns` Collection already has an entry for the command being used. If this is not the case, you can add a new entry, where the value is initialized as an empty Collection. Next, create the following variables:

1. `now`: The current timestamp.
2. `timestamps`: A reference to the Collection of user ids and timestamp key/value pairs for the triggered command.
3. `cooldownAmount`: The specified cooldown for the command, converted to milliseconds for straightforward calculation. If none is specified, this defaults to three seconds.

If the user has already used this command in this session, get the timestamp, calculate the expiration time and inform the user of the amount of time they need to wait before using this command again. Note the use of `return` statement here, causing the code below this snippet to execute only if the user has not used this command in this session or the wait has already expired.

Continuing with your current setup, this is the complete `if` statement:

```js {2,4-7}
if (timestamps.has(interaction.user.id)) {
	const expirationTime = timestamps.get(interaction.user.id) + cooldownAmount;
	if (now < expirationTime) {

> 		const timeLeft = Math.round(expirationTime / 1000);

Reviewer 2: Still its not *time left*
because its just unix timestamp

Author: unix timestamp of the time left, no?

Reviewer 3: It’s the (absolute) timestamp of when the cooldown expires, not a relative duration. So `expirationTimestamp` would be more appropriate here

# Cooldowns

Spam is something you generally want to avoid, especially if one of your commands require calls to other APIs or takes a bit of time to build/send.

::: tip
This section assumes you followed the [Command Handling](/creating-your-bot/command-handling.md) part.
:::

First, add a cooldown property to your command. This will determine how long the user would have to wait (in seconds) before using the command again.

```js {4}
const { SlashCommandBuilder } = require('discord.js');

module.exports = {
	cooldown: 5,
	data: new SlashCommandBuilder()
		.setName('ping')
		.setDescription('Replies with Pong!'),
	async execute(interaction) {
		// ...
	},
};
```

> In your main file, initialize a [Collection](/additional-info/collections.md) to store cooldowns of commands

Reviewer 4: ```suggestion
In your main file, initialize a [Collection](/additional-info/collections.md) to store cooldowns of commands:
```

In your main file, add the following code:

```js {1,3-5,7-10,12-14}
const { cooldowns } = client;

if (!cooldowns.has(command.data.name)) {
	cooldowns.set(command.data.name, new Collection());
}

const now = Date.now();
const timestamps = cooldowns.get(command.data.name);
const defaultCooldownDuration = 3;
const cooldownAmount = (command.cooldown ?? defaultCooldownDuration) * 1000;

if (timestamps.has(interaction.user.id)) {
	// ...
}

try {
	// ...
} catch (error) {
	// ...
}
```

You check if the `cooldowns` Collection already has an entry for the command being used. If this is not the case, you can add a new entry, where the value is initialized as an empty Collection. Next, create the following variables:

1. `now`: The current timestamp.
2. `timestamps`: A reference to the Collection of user ids and timestamp key/value pairs for the triggered command.
3. `cooldownAmount`: The specified cooldown for the command, converted to milliseconds for straightforward calculation. If none is specified, this defaults to three seconds.

> If the user has already used this command in this session, get the timestamp, calculate the expiration time and inform the user of the amount of time they need to wait before using this command again. Note the use of `return` statement here, causing the code below this snippet to execute only if the user has not used this command in this session or the wait has already expired.

Reviewer 4: ```suggestion
If the user has already used this command in this session, get the timestamp, calculate the expiration time, and inform the user of the amount of time they need to wait before using this command again. Note the use of the `return` statement here, causing the code below this snippet to execute only if the user has not used this command in this session or the wait has already expired.
```

Continuing with your current setup, this is the complete `if` statement:

```js {2-7}
if (timestamps.has(interaction.user.id)) {
	const expirationTime = timestamps.get(interaction.user.id) + cooldownAmount;

	if (now < expirationTime) {
		const expiredTimestamp = Math.round(expirationTime / 1000);
		return interaction.reply({ content: `Please wait <t:${expiredTimestamp}:R> more second(s) before reusing the \`${command.data.name}\` command.`, ephemeral: true });
	}
}
```

> Since the `timestamps` Collection has the user's id as the key, you can use the `get()`method on it to get the value and sum it up with the `cooldownAmount` variable to get the correct expiration timestamp and further check to see if it's expired or not.

Reviewer: ```suggestion
Since the `timestamps` Collection has the user's id as the key, you can use the `get()` method on it to get the value and sum it up with the `cooldownAmount` variable to get the correct expiration timestamp and further check to see if it's expired or not.
```

> Since the `timestamps` Collection has the user's id as the key, you can use the `get()`method on it to get the value and sum it up with the `cooldownAmount` variable to get the correct expiration timestamp and further check to see if it's expired or not.

Reviewer 4: ```suggestion
Since the `timestamps` Collection has the user's id as the key, you can use the `get()` method on it to get the value and sum it up with the `cooldownAmount` variable to get the correct expiration timestamp and further check to see if it's expired or not.
```

> The previous user check serves as a precaution incase the user leaves the guild. You can now use the `setTimeout` method, which will allow you to execute a function after a specified amount of time and remove the timeout.

Reviewer 4: ```suggestion
The previous user check serves as a precaution in case the user leaves the guild. You can now use the `setTimeout` method, which will allow you to execute a function after a specified amount of time and remove the timeout.
```
